storm/src/main/storm/mesos/resources/ResourceEntries.java

```diff
@@ -0,0 +1,111 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package storm.mesos.resources;
+
+public class ResourceEntries {
+  public static final class ScalarResourceEntry implements ResourceEntry<Double> {
+    private ReservationType reservationType;
+    private Double value;
+
+    public ReservationType getReservationType() {
+      return reservationType;
+    }
+
+    public ScalarResourceEntry(Double value) {
+      this.value = value;
+    }
+
+    public ScalarResourceEntry(ReservationType reservationType, Double value) {
+      this.reservationType = reservationType;
+      this.value = value;
+    }
+
+    public Double getValue() {
+      return value;
+    }
+
+    @Override
+    public ScalarResourceEntry add(ResourceEntry<Double> resourceEntry) {
+      ScalarResourceEntry scalarResourceEntry = (ScalarResourceEntry) resourceEntry;
+      this.value += scalarResourceEntry.getValue();
+      return this;
+    }
+
+    @Override
+    public ScalarResourceEntry remove(ResourceEntry<Double> resourceEntry) {
+      ScalarResourceEntry scalarResourceEntry = (ScalarResourceEntry) resourceEntry;
+      this.value -= scalarResourceEntry.getValue();
+      return this;
+    }
+  }
+
+  public static final class RangeResourceEntry implements ResourceEntry<Long> {
+    private ReservationType reservationType;
+    private Long begin;
+    private Long end;
+
+    public RangeResourceEntry(Long begin, Long end) {
+      this.reservationType = reservationType;
+      this.begin = begin;
+      this.end = end;
+    }
+
+    public RangeResourceEntry(ReservationType reservationType, Long begin, Long end) {
+      this.reservationType = reservationType;
+      this.begin = begin;
+      this.end = end;
+    }
+
+    public Long getBegin() {
+      return begin;
+    }
+
+    public Long getEnd() {
+      return end;
+    }
+
+    public ReservationType getReservationType() {
+      return reservationType;
+    }
+
+    @Override
+    public RangeResourceEntry add(ResourceEntry<Long> resourceEntry) {
```

Comment:
For this method, if we have a range say, [u,v] and we want to add [w,x], the only cases we can actually do this are when the result is [u,x] (when v+1 = w) or [w,v] (when x+1 = u). Unless I'm misunderstanding those are the only options because we can't have two separate ranges, just one range.

If this is the case, we need to have the following logic instead of what is there:
```
if (end+1 != rangeResource.getBegin() && begin != rangeResource.end()+1) {
    //incompatible addition, throw an exception?
}
if (end < rangeResourceEntry.getBegin()) {
    end = rangeResourceEntry.getEnd();
} else if (begin > rangeResourceEntry.getEnd()) {
    begin = rangeResourceEntry.getBegin();
}
``` 

If that's not the case, can you explain how this is supposed to work? (Either way, I think the logic below is flawed because both have to do with using only the beginning of the `rangeResourceEntry`).